Feature/planning v2

orthagh opened the pull request:
| Q             | A
| ------------- | ---
| Bug fix?      | yes
| New feature?  | yes
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes

See #5917 

Changes proposed in this PR:

- [X] Add scheduler view
- [X] clean code for removed callback/options in fullcalendar v4
- [X] move code for controlling displayed events in the planning js file
- [X] fix some issues not found before merging #6025 

Preview:

![image](https://user-images.githubusercontent.com/418844/61541402-e0a3bf00-aa3f-11e9-8d35-cf43c0d6dd9b.png)

**Important**: scheduler view needs the following plugin installed in GLPI:
[advancedplanning.zip](https://github.com/glpi-project/glpi/files/3397093/advancedplanning.zip)

[Fullcalendar Scheduler](https://github.com/fullcalendar/fullcalendar-scheduler) library is released on GPL **v3** licence incompatible with GPL **v2** of GLPI.
I don't have any idea how to achieve more simpler delivery for this part.

cedric-anne commented on inc/planning.class.php:
```diff
@@ -1637,14 +1573,45 @@ static function updateEventTimes($params = []) {
+                  $new_actor_itemtype = strtolower($params['new_actor_itemtype']);
+
+                  // reminders don't have group assignement for planning
+                  if (!($new_actor_itemtype === 'Group'
```
`$new_actor_itemtype` is lowered.
```suggestion
                  if (!($new_actor_itemtype === 'group'
```

cedric-anne commented:
Some remarks :
 - resources view displays lines even for users/groups that are not "active" (not checked),
 - moving a project task to another user/group does not work;

orthagh commented:
Fixes for review done

orthagh commented:
I'll take the time in this pr to add more ux features:

- Scroll by default to today/now in resources or week view
- Set cookie for the last view viewed
- clone events by ctrl+drag

cedric-anne commented on inc/planningevent.class.php:
```diff
@@ -0,0 +1,417 @@
+            $this->fields["begin"],
+            $this->fields["end"],
+            [
+               'Reminder' => [$this->fields['id']]
```
```suggestion
               $this->getType() => [$this->fields['id']]
```

cedric-anne commented on inc/planningevent.class.php:
```diff
@@ -0,0 +1,417 @@
+            $this->fields["begin"],
+            $this->fields["end"],
+            [
+               self::getType() => [$this->fields['id']]
```
```suggestion
               $this->getType() => [$this->fields['id']]
```

cedric-anne commented on inc/planningeventcategory.class.php:
```diff
@@ -0,0 +1,43 @@
+
+
+   static function getTypeName($nb = 0) {
+      return _n('User category', 'User categories', $nb);
```
```suggestion
      return _n('Event category', 'Event categories', $nb);
```

cedric-anne commented on js/planning.js:
```diff
@@ -324,16 +378,227 @@ var GLPIPlanning = {
    },
 
+   refresh: function() {
+      GLPIPlanning.calendar.refetchResources();
```
Function does not exists if resource plugin is not loaded.
```suggestion
      if (typeof(GLPIPlanning.calendar.refetchResources) == 'function') {
         GLPIPlanning.calendar.refetchResources();
      }
```

orthagh commented on inc/define.php:
```diff
@@ -385,8 +385,8 @@
-                                                  'TicketTask', 'ProjectTask'];
-$CFG_GLPI['planning_add_types']           = ['Reminder'];
+                                             'TicketTask', 'ProjectTask', 'PlanningExternalEvent'];
+$CFG_GLPI['planning_add_types']           = ['PlanningExternalEvent'];
```
yes

orthagh commented on inc/group_user.class.php:
```diff
@@ -789,4 +789,101 @@ protected static function getListForItemParams(CommonDBTM $item, $noent = false)
+
+         // save the planning completed to db
+         $json_plannings = exportArrayToDB($plannings);
+         $stmt->bind_param('si', $json_plannings, $users_id);
```
if you have suggestion for this point, i don't atm how to do otherwise

cedric-anne commented on inc/planningexternalevent.class.php:
```diff
@@ -0,0 +1,164 @@
+         ]);
+      }
+      echo "</td></tr>";
+
```
CS
```suggestion
```

cedric-anne commented:
I just noticed that:
 - there are no search options for exetnal events, so displaying more than name in list is impossible,
 - edit form of a planning event is broken (display seems to fail on rrule field).

orthagh commented:
because it's currently a draft not a finished implementation. Rrule is barely started

cedric-anne commented on inc/planningexternalevent.class.php:
```diff
@@ -0,0 +1,163 @@
+
+      echo "<tr class='tab_bg_2'><td  colspan='2'>".__('Repeat')."</td>";
+      echo "<td>";
+      echo self::showRepetitionForm($this->fields['rrule']);
```
If field is null in DB, passed value is not a string so it triggers a type error.
```suggestion
      echo self::showRepetitionForm($this->fields['rrule'] ?? '{}');
```

orthagh commented:
All features done, ready for final review @cedric-anne

cedric-anne commented:
I found some issues:

- Recurring events are not only visible in month/week/day planning views if their first occurence is not within displayed time frame.

- Moving an instance of recurring event can have unexpected result:
  - It will define a new begin date for all occurences, so , most of the time, it will make previous occurences disapear.
  - If new begin date does not match rrule criteria, event can disapear (for example, moving on tuesday an event made to be repeated only on mondays).

orthagh commented:
@Canne, Issues fixed

For 2nd one, i abort the drag/resize events or reccuring event, i didn't found any easy way to handle them. Edition is still possible by the way.

canne commented:
Hi @orthagh, I keep receiving messages regularly from this project under your name. I noticed that there is cedric-anne above, are you mixing it with my ID, @canne, by any chance? Thank you, Canne.

orthagh commented:
Yes, sorry, it's the second time you say it to me.
I'll try to be more cautious

cedric-anne commented:
Some remarks.

1. In timeline week view, events (background or not) uses category color. In other views, events uses user color. I did not understand why there was a difference at first glance.

2. I do not see background events in monthly view.

3. When adding multiple items, the message that indicates that item is created refers to previously created item (instead of last created item)

4. Looks like recurring events are disturbing business checks for a user. See following image, I created an event on an empty timeslot, but I had a message telling me that my recurring event is already using this time slot.
![image](https://user-images.githubusercontent.com/33253653/65681132-a0247b00-e058-11e9-9be7-19772c4c5702.png)

orthagh commented:
1- yes, specific view btw, i added a comment for this
2- yes, only `allday` events are displayed when set as background in month view
3- messages after redirected fixed
4- fixed by the last commit, i explode recurrent event with https://github.com/rlanvin/php-rrule

i added additional fixes :

- limit recurrent events to one iteration (the next one following current date) in list view to keep performance ok (zimbra do as well)
- added icon (and title attribute) to distinguish recurrent event
- added a replacement for rrulejs change for byweekday key (byday in rrule.js)
- added a new right for managing backgound event (override update right).

Should be ok now, i hope a merge :p

cedric-anne commented on inc/planning.class.php:
```diff
@@ -1524,30 +1538,105 @@ static function constructEventsArray($options = []) {
+
+         // manage reccurent events
+         if (isset($event['rrule']) && count($event['rrule'])) {
+            $rrule = array_merge($event['rrule'], [
```
I had to do this to prevent having an exception:

```php
            $startDt = new \DateTime($new_event['start']);
            $rrule = array_merge($event['rrule'], [
               'dtstart' => $startDt->format('Ymd\THis\Z')
            ]);
```

Exception was: 
```
<br />
<b>Fatal error</b>:  Uncaught RuntimeException: IntlDateFormatter::create() failed. Error Code: 1 &quot;datefmt_create: no such time zone: 'GMT+00:00': U_ILLEGAL_ARGUMENT_ERROR&quot; (this should not happen, please open a bug report!) in /var/www/glpi/vendor/rlanvin/php-rrule/src/RRule.php:2236
Stack trace:
#0 /var/www/glpi/inc/planning.class.php(1588): RRule\RRule-&gt;humanReadable()
#1 /var/www/glpi/ajax/planning.php(43): Planning::constructEventsArray(Array)
#2 {main}
  thrown in <b>/var/www/glpi/vendor/rlanvin/php-rrule/src/RRule.php</b> on line <b>2236</b><br />
```

cedric-anne commented:
> 1. I have the `Debug mode has been activated` message shown each time I create an event from the planning view.
Cannot reproduce, maybe a problem with session.

cedric-anne commented:
> 2. Recurring event start date is always the current datetime in month/week/day views.

Fixed too.
